### ISSUE TITLE
Log tracking

### ISSUE THREAD

**Issue by let4be** [lavab/web#279]
We need some way to collect logs from the application to be able to debug issues, without logs we're like without hands...

Also we're heavily using js/css transpiling so we also need to deal with source maps:
https://trello.com/c/dDCrHeid/136-source-maps-on-production

This one looks promising: https://getsentry.com/welcome/

@andreis, @pzduniak , @flixi ideas are welcomed

**Comment by let4be** [comment]
Other alternatives to consider:
https://errorception.com/
https://airbrake.io/
http://www.exceptionhub.com/

**Comment by flixi** [comment]
Status?

**Comment by flixi** [comment]
We need to make sure that the user's integrity is not compromised though.

**Comment by andreis** [comment]
We're gonna have a log management system on openstack, the question is how to send logs from the front end.

I suggest a "Report error" button that can be enabled/disabled in the Settings. Clicking the button warns the user that we'll send anonymized console logs to our servers, then proceeds to do so.

**Comment by let4be** [comment]
thing is: to be able to reproduce some kind of errors only error message and stacktrace is not enough, full logs required

**Comment by let4be** [comment]
also one another thing is making me sad: we still do not have proper stacktraces, 
stacktrace + transpiling we use + source maps = mumbo jumbo

I have no idea how to solve this issue, I've tried all possible ways - nothing helps, most likely this is browser related thing

It stills possible to locate the code in transpiled source but it takes more time

**Comment by andreis** [comment]
Helps?

* http://stackoverflow.com/questions/25827234/how-to-debug-javascript-promises
* https://github.com/soareschen/es6-promise-debugging/blob/master/README.md

**Comment by andreis** [comment]
Most likely people will report bugs in `prod`, so it might be an idea to switch the output of the logging from `console` to something that can be read at runtime (i.e. just store it in an array or something).

**Comment by flixi** [comment]
A few users will also be allowed on `staging`so it might not be only `prod` who's going to report bugs.

**Comment by flixi** [comment]
I like your ideas @andreis. However we need the ability and make sure to inform the user of things such as personal information may be transmitted. But we will keep all the info personal and delete the logs as soon as the issue seems to be fixed.

**Comment by flixi** [comment]
Status @andreis ?

**Comment by let4be** [comment]
I'm going to work on this soon, this is one of the most important things we need before we go into wide use of lavaboom 

meanwhile any ideas how and when we should ask user about logs collection and submission? 

@vLooz @flixi

**Comment by let4be** [comment]
we also need to:
1. send version number within the logs (x.y.z, the one from manifest.json)
1. save source maps of each github release so we could always match x.y.z logs with particular source maps server-side

**Comment by let4be** [comment]
so... There should be 
- a button/slider to turn on and turn off log tracking(asking during signup and available under settings)
- a button to report an issue, probably with some kind of form where user could briefly describe what happened. What do you think?

Any design guidelines where better to put this? 
@flixi @vLooz

**Comment by flixi** [comment]
No, but we might want to integrate this into a bugzilla installation.

**Comment by let4be** [comment]
So this turned out to be not so simple task as I thought initially because of very limited capabilites of Session Storage, and I totally agree with @andreis - Session Storage backup required for proper work of this feature as page refresh is a common user's reaction to a sudden bug and in this case Session Storage will hold out data.

So far, we have to:
- store logs in json as session storage supports only strings
- store logs in chunks as we can't simply concatenate a string stored in session storage performance efficient(i.e. lavab-logs-chunk-0, lavab-logs-chunk-1, etc...) this would allow to significantly reduce the amount of wasted cpu cycles on saving, json encoding and decoding. Especially profitable with big logs which is anyway limited to 5MB(session/local storage cross browser limit)
- don't write immediately after each intercepted log call, but instead serialize and dump data after some timeout interval(set to 1 sec currently)
- use compression to increase the amount of stored logs(http://pieroxy.net/blog/pages/lz-string/index.html)

This is already almost implemented

**Comment by andreis** [comment]
Most likely we'll need an API endpoint for saving logs. Either `api.lavaboom.com/logs` or `logs.lavaboom.com`

**Comment by let4be** [comment]
Okay, core logic almost ready from my side.
What's left to do:
- conditionally turn and off logs/errors collection
- clear stored logs
- bind to report api endpoint
- integrate into UI, open to ideas and suggestions. 
@andreis @flixi @vLooz how this should look like from the user perspective and what about bugzilla?

Here is the latest specification - source maps and example of report in json format
https://www.dropbox.com/s/ito2ts4wdbr5n43/logs.7z?dl=0
(wanted to attach on github but: "Unfortunately, we don't support that file type")

**Comment by andreis** [comment]
:+1:

**Comment by let4be** [comment]
So the question stills actual! How this should look like from UX point and what flow are we going to choose?

**Comment by vLooz** [comment]
I have to admit that I only understand a fraction of all this technical jargon ;-) 
A button in settings to turn ON/OFF automatic bug reports + the necessary info what this means for privacy and a manual bug-report Icon I can design though.

How would you guys like the idea if the latter (bug-report button) would be above the log out button in the left panel? Of course only in the BETA accounts

**Comment by let4be** [comment]
@vLooz there are 2 steps:
1. background logs/error collection and persistence between page refresh
2. error reporting that will include: subject, description, probably screenshot and other related information(to be discussed how and what) + all collected and persisted(to browser session storage) logs && errors

so there won't be any automatic bug reports . The thing that can be done automatically is logs/errors collection, here we can go for ON/OFF toggle.

The other part and how I see it:
if user meets some issue he can press "Report bug" button where he will be able to describe his issue and the system will automatically register this issue via our API with the logs attached.

Here goes the question about bugzilla... are we going to use it and if so how should it be integrated? @andreis @flixi

**Comment by let4be** [comment]
Actually I would implement second step via our super duper compose screen ;)
User presses "Report bug", compose screen opens with prefilled to, subject, body and attached logs

What do you think @vLooz @andreis @flixi ?

**Comment by vLooz** [comment]
This are 2 proposals for the look of the "report a bug function"
 
![bildschirmfoto 2015-05-07 um 15 54 10](https://cloud.githubusercontent.com/assets/11177812/7516716/5bc15900-f4d1-11e4-8242-f7d08e1ad641.png)

**Comment by andreis** [comment]
First one, totally.
@vLooz

**Comment by andreis** [comment]
@let4be: We'll definitely use Bugzilla or something like it (Trac?). Most likely hosted at https://lavab.wtf

**Comment by let4be** [comment]
Sounds good

**Comment by vLooz** [comment]
I like the first one better, too. 
The icon is in the updated icon font, look for "attention" @piggyslasher,

**Comment by vLooz** [comment]
This is how the toggle switch field in settings could look like:

![279 activate auto-logs](https://cloud.githubusercontent.com/assets/11177812/7520452/25319ba0-f4e7-11e4-85ad-002805821f74.png)

**Comment by vLooz** [comment]
By the way: Is there any way to obfuscate the content of the logged emails? I don't feel very good with the idea, that we can actually read our users emails. @andreis @pzduniak

**Comment by let4be** [comment]
@vLooz There are 2 groups of issues,
one class of them requires access to full email body that caused the issue and another class can be debugged well without having this information

I suggest to give user a possibility to remove email bodies before doing a report, with a proper warning that it may make issue resolution impossible

**Comment by vLooz** [comment]
Ok, thx for the info @let4be 
But coulnd't the first class also work with obfuscated mail content? I mean, it would still get the full email body, but with different (unreadable) mail content. Would that be an option?

**Comment by let4be** [comment]
unfortunately no., 
we perform extensive email post processing in order to display email:
- replace links to make them clickable
- replace emails to make them clickable and add context menu to them
- decrypt other embedded openpgp envelopes(implemented today)
+ perform email body sanitization again malicious content

to debug all of this full original email body is must have

**Comment by vLooz** [comment]
Humm... I still don't like it. The thing is, zero knowledge is one of our main, if not THE main user benefit we advertise. It's what we sell to our users, it's what we claim to REALLY do different compared to all of those competitors who promise this, but in the background don't keep. In my opinion we just CAN'T read out clients mail content, otherwise our whole story collapses and we loose all the credibility, we've earned so far. Just my pov. What do you think about this @flixi?

**Comment by let4be** [comment]
We can make this beta-only feature with a proper warning, or we may not log such content at all thus we won't be able to debug some kind of issues at all.

**Comment by let4be** [comment]
Issues that require full email body to be resolve take relatively small fraction from all other issues

**Comment by vLooz** [comment]
Would there be a possibility to display the mail content, to be send within the log file, before the mail is send? So that the user can decide case by case if she wants to share this content or not?

**Comment by vLooz** [comment]
what do you think about this @andreis ?

**Comment by let4be** [comment]
I prefer having those logs totally removed, we already have enough complexities in this system

**Comment by vLooz** [comment]
From a user privacy point of view this would be the best option, imo.

**Comment by andreis** [comment]
There is a simple solution to this. Have the normal log collection (background or on demand) obfuscate private data, and have a *secret* keybinding for sending private data with logs.

We can use that command, and also share it with users who want to help fix an eventual issue.

**Comment by flixi** [comment]
Agreed with @let4be let's remove those logs with email contents in them.

**Comment by flixi** [comment]
Don't make it too complicated though @andreis with keybinding etc.

**Comment by let4be** [comment]
Most of stuff is ready here and I just need to integrate logs/sourcemaps submitting, will handle this Monday

**Comment by andreis** [comment]
Advice from Github https://news.ycombinator.com/item?id=9570519

cc @pzduniak @let4be

**Comment by pzduniak** [comment]
hn is dead for me now
20 maj 2015 09:03 "Andrei Simionescu" <notifications@github.com> napisał(a):

**Comment by andreis** [comment]
But Piotr is alive!

**Comment by let4be** [comment]
api endpoint wanted

**Comment by pzduniak** [comment]
Done! API endpoint is described on #dev-web on Slack.

**Comment by vLooz** [comment]
Hey @piggyslasher @let4be, has there been any decision about how to implement this button?

![bildschirmfoto 2015-05-07 um 15 54 10 kopie](https://cloud.githubusercontent.com/assets/11177812/8474640/cdd76028-20b1-11e5-8900-1ac0433c312b.png)

**Comment by andreis** [comment]
I wish @pzduniak didn't use Slack to describe said functionality...
